python/ray/serve/_private/application_state.py

```diff
@@ -148,22 +197,47 @@ def delete_application(self, name: str):
             return
         self._application_states[name].delete()
 
-    def deploy_application(self, name: str, deployment_args: List[Dict]):
+    def deploy_application(self, name: str, deployment_args: List[Dict]) -> bool:
         """Deploy single application
 
         Args:
             name: application name
             deployment_args: deployment args
+        Returns:
+            whether the application is deployed successfully or not.
         """
-        if name in self._application_states:
-            self._application_states[name].deployment_params = deployment_args
-        else:
+
+        # Make sure route_prefix is not being used by other application.
+        all_route_prefixes: Dict[str, str] = {
+            self._application_states[app_name].route_prefix: app_name
+            for app_name in self._application_states
+        }
+        for deploy_param in deployment_args:
+            if "route_prefix" in deploy_param:
+                deploy_app_prefix = deploy_param["route_prefix"]
+                if (
+                    deploy_app_prefix
+                    and deploy_app_prefix in all_route_prefixes
+                    and name != all_route_prefixes[deploy_app_prefix]
+                ):
+                    logger.error(
+                        f"Prefix {deploy_app_prefix} is being used by application "
+                        f"{all_route_prefixes[deploy_app_prefix]},"
+                        f" failed to deploy application {name}."
+                    )
+                    return False
```

Comment:
this should probably raise an exception or return an error message to the client

for example in the REST API we need to return this somehow